TLM/TLM/Util/RoundaboutMassEdit.cs

```diff
@@ -359,30 +386,67 @@ public class RoundaboutMassEdit {
+            foreach (ushort segmentId in segList)
+                record.AddCompleteSegment(segmentId);
+
+            // Add each minor road.
+            foreach (ushort nodeId in record.NodeIDs.ToArray()) {
+                ref NetNode node = ref nodeId.ToNode();
+                if (node.CountSegments() < 3) continue;
+                for (int i = 0; i < 8; ++i) {
+                    ushort segmentId = node.GetSegment(i);
+                    if (segmentId == 0)
+                        continue;
+                    if (record.SegmentIDs.Contains(segmentId))
+                        continue;
+
+                    record.AddSegmentAndNodes(segmentId);
                 }
-                return true;
-            });
+            }
+
+            record.Record();
+            return record;
         }
 
         /// <summary>
-        /// Clears all rules put by RoundAboutMassEdit.FixRoundabout()
+        /// Calculates Raduis of a curved segment assuming it is part of a circle.
         /// </summary>
-        /// <param name="segList"></param>
-        public void ClearRoundabout(List<ushort> segList) {
-            foreach (ushort segmentId in segList) {
-                foreach (bool startNode in Constants.ALL_BOOL) {
-                    ushort nodeId = netService.GetSegmentNodeId(segmentId, startNode);
-                    ClearNode(nodeId);
-                }
+        internal static float CalculateRadius(ref NetSegment segment) {
+            // TDOO: to calculate maximum curviture for eleptical roundabout, cut the bezier in 10 portions
+            // and then find the bezier with minimum raduis.
+            Vector2 startDir = VectorUtils.XZ(segment.m_startDirection);
+            Vector2 endDir = VectorUtils.XZ(segment.m_endDirection);
+            Vector2 startPos = VectorUtils.XZ(segment.m_startNode.ToNode().m_position);
+            Vector2 endPos = VectorUtils.XZ(segment.m_endNode.ToNode().m_position);
+            float dot = Vector2.Dot(startDir, -endDir);
+            float len = (startPos - endPos).magnitude;
+            float r = len / Mathf.Sqrt(2 - 2 * dot); // see https://github.com/CitiesSkylinesMods/TMPE/issues/793#issuecomment-616351792
+            return r;
+        }
+
+
+        /// <summary>
+        /// calculates realisitic speed limit of input curved segment assuming it is part of a circle.
+        /// minimum speed is 10kmph.
+        /// </summary>
+        /// <returns>Null if segment is straight, otherwise if successful it retunrs calcualted speed.</returns>
+        private static SpeedValue? CalculatePreferedSpeed(ushort segmentId) {
+            float r = CalculateRadius(ref segmentId.ToSegment());
+            float kmph = 11.3f * Mathf.Sqrt(r); // see https://github.com/CitiesSkylinesMods/TMPE/issues/793#issue-589462235
+            Log._Debug($"CalculatePreferedSpeed radius:{r} -> kmph:{kmph}");
+            if (float.IsNaN(kmph) || float.IsInfinity(kmph) || kmph < 1) {
```

Comment:
use 1f, i have a feeling that this would compile to `int 1`
same below: 10f instead of 10